includes/profile.inc

```diff
@@ -50,60 +50,117 @@ function cm_tools_module_disable($modules) {
 }
 
 /**
- * Apply given callback to the entities returned by the given EntityFieldQuery.
+ * Apply given work_callback to the entities supplied.
  *
  * This function is designed to work in the context of a Drupal update hook, it
  * accepts the sandbox that is passed to the update hook as its first parameter.
- * The results of the query, and the state of the iterations are stored in the
- * sandbox.
+ * The entities to iterate over and the state of the iterations are stored in
+ * the sandbox.
  *
  * @param array $sandbox
  *   The sandbox passed to the drupal hook hook that this function is called
  *   from.
- * @param \EntityFieldQuery $query
- *   The EntityFieldQuery to execute and iterate over the results of.
- * @param callable $callback
- *   The callback to apply to each entity returned from the $query. The callback
- *   is called with two arguments:
+ * @param callable $entity_list_callback
+ *   A callback that returns an associative array keyed by entity type where the
+ *   values are arrays of entity IDs to iterate over. This will be called when
+ *   needed to produce the list of entities to iterate over.
+ * @param callable $work_callback
+ *   The callback to apply to each entity returned from the
+ *   $entity_list_callback. The callback is called with two arguments:
  *   - The entity type of the entity in the second argument
  *   - The loaded entity.
  * @param int $entities_per_pass
  *   The number of entities to load and process in each batch iteration.
+ *
+ * Example usage:
+ * @code
+ * function node_update_7017(&$sandbox) {
+ *   $entity_list_callback = function () {
+ *     $select = db_select('node', 'n')
+ *       ->condition('n.type', 'article')
+ *       ->fields('n', array('nid'));
+ *     $result = $select->execute()->fetchCol();
+ *     $entities = array();
+ *     if (!empty($result)) {
+ *       $entities['node'] = $result;
+ *     }
+ *     return $entities;
+ *   };
+ *   $work_callback = function ($entity_type, $entity) {
+ *     node_save($entity);
+ *   };
+ *   cm_tools_include('profile');
+ *   // Load and save all article nodes.
+ *   cm_tools_update_entities_walk($sandbox, $entity_list_callback, $work_callback);
```

Comment:
Since this code example will become the template for dozens of update hooks, here are some thoughts:

**Question (DX):** Why not pass `$entity_list` as the _result_ rather than as a callback? It gets executed immediately anyway, and I wonder if that would give us less confusing code in the update hook?

**Question (DX)**: My gut tells me that it seems neater to specify `$work_callback` inline, rather than via a variable.

Reply:
@jamsilver I agree with your second question. Not your first :-) (Meatspace?)

Reply:
Aaaaand @anotherjames has answered that by pointing out that we _only_ want to run it on the first run and not subsequent runs so the callback is the right approach 👍

Reply:
It's an example, you may write your actual code as you wish.
But...honestly, no, usually the actual callback is several lines long so I disagree that it's neater to have it inline in the middle of the function call.

I've just written one for a project that has several conditionals and a try/catch block etc. _That_ inline in the middle of a function call is not nice imho. But you are free to inline when writing your own implementations.